File: backend/src/flows.ts
import { z } from 'zod';

const Node = z.object({
  id: z.string().regex(/^[A-Za-z0-9_-]{1,64}$/),
  type: z.enum(['answer','play','menu','queue','ai_agent','knowledge','meeting','webhook','voicemail','hangup']),
  config: z.record(z.string(), z.unknown()).default({})
});
const Edge = z.object({source:z.string(),target:z.string(),outcome:z.string().default('default')});
export const Flow = z.object({name:z.string().min(3).max(120),entryNodeId:z.string(),nodes:z.array(Node).min(1).max(200),edges:z.array(Edge).max(500)});

export function validateFlow(input: unknown) {
  const flow = Flow.parse(input); const ids = new Set(flow.nodes.map(n=>n.id));
  if (ids.size !== flow.nodes.length) throw new Error('node IDs must be unique');
  if (!ids.has(flow.entryNodeId)) throw new Error('entry node does not exist');
  for (const edge of flow.edges) if (!ids.has(edge.source)||!ids.has(edge.target)) throw new Error('edge references unknown node');
  const reached = new Set([flow.entryNodeId]); let changed = true;
  while(changed){changed=false;for(const e of flow.edges)if(reached.has(e.source)&&!reached.has(e.target)){reached.add(e.target);changed=true;}}
  const unreachable=[...ids].filter(id=>!reached.has(id)); if(unreachable.length) throw new Error(`unreachable nodes: ${unreachable.join(', ')}`);
  if(!flow.nodes.some(n=>n.type==='hangup')) throw new Error('flow requires a hangup node');
  return flow;
}

File: services/control-plane/app/flows.py
from enum import StrEnum

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field, model_validator

from .security import Principal, require_tenant

router = APIRouter(prefix="/v1/flows", tags=["flows"])


class NodeType(StrEnum):
    answer = "answer"
    play = "play"
    menu = "menu"
    queue = "queue"
    webhook = "webhook"
    voicemail = "voicemail"
    hangup = "hangup"


class FlowNode(BaseModel):
    id: str = Field(pattern=r"^[A-Za-z0-9_-]{1,64}$")
    type: NodeType
    config: dict = Field(default_factory=dict)


class FlowEdge(BaseModel):
    source: str
    target: str
    outcome: str = "default"


class FlowDefinition(BaseModel):
    name: str = Field(min_length=3, max_length=120)
    entry_node_id: str
    nodes: list[FlowNode] = Field(min_length=1, max_length=200)
    edges: list[FlowEdge] = Field(default_factory=list, max_length=500)

    @model_validator(mode="after")
    def validate_graph(self):
        ids = [node.id for node in self.nodes]
        if len(ids) != len(set(ids)):
            raise ValueError("node IDs must be unique")
        known = set(ids)
        if self.entry_node_id not in known:
            raise ValueError("entry node does not exist")
        for edge in self.edges:
            if edge.source not in known or edge.target not in known:
                raise ValueError("edge references an unknown node")
        reachable = {self.entry_node_id}
        changed = True
        while changed:
            changed = False
            for edge in self.edges:
                if edge.source in reachable and edge.target not in reachable:
                    reachable.add(edge.target)
                    changed = True
        unreachable = known - reachable
        if unreachable:
            raise ValueError(f"unreachable nodes: {', '.join(sorted(unreachable))}")
        if not any(node.type == NodeType.hangup for node in self.nodes):
            raise ValueError("flow requires a hangup node")
        return self


class ValidationResult(BaseModel):
    valid: bool
    node_count: int
    edge_count: int


@router.post("/validate", response_model=ValidationResult)
def validate_flow(
    definition: FlowDefinition,
    principal: Principal = Depends(require_tenant),
) -> ValidationResult:
    return ValidationResult(
        valid=True,
        node_count=len(definition.nodes),
        edge_count=len(definition.edges),
    )
